### PR TITLE
provides a module initialization method

### PR DESCRIPTION
**Pull request by baichuan3**
reslove https://github.com/rooch-network/rooch/issues/63

since the Move runtime does not to know about new packages created,
the first deployment contract and the upgrade contract need to be handled separately

The initializer function must have the following properties in order to be executed at publication:
- Name init
- Single parameter of &mut TxContext type
- No return values
- Private & no Entry

In addition, the Ocject information is finally output through the event, and the event has not been realized yet, so the log output is added first, so that the process can be connected through the cli.

### CODE REVIEW

File: moveos/moveos/src/moveos.rs [hunk 1]
```diff
@@ -175,6 +193,144 @@ impl MoveOS {
...
+            assert!(
+                result.return_values.is_empty(),
+                "init should not have return values"
+            )
+        }
+
+        Ok(())
+    }
+
+    fn execute_move_call_bypass_visibility<S>(
```
Comment (jolestar) [inline moveos/moveos/src/moveos.rs:305]: this function conflict with #143 , need to resolve

File: moveos/moveos/src/moveos.rs [hunk 2]
```diff
@@ -175,6 +193,144 @@ impl MoveOS {
...
+                &module_id,
+                INIT_FN_NAME,
+                vec![],
+                args,
+            )?;
+
+            assert!(
+                result.return_values.is_empty(),
+                "init should not have return values"
+            )
```
Comment (jolestar) [inline moveos/moveos/src/moveos.rs:299]: use bail! to raise the error, assert will panic the VM.

File: moveos/moveos/src/moveos.rs [hunk 3]
```diff
@@ -175,6 +193,144 @@ impl MoveOS {
...
+            !modules.is_empty(),
+            "input checker ensures package is not empty"
+        );
+
+        Ok(modules)
+    }
+
+    /// The initializer function must have the following properties in order to be executed at publication:
+    /// - Name init
+    /// - Single parameter of &mut TxContext type
```
Comment (jolestar) [inline moveos/moveos/src/moveos.rs:213]: Is support `signer`? if not, can not call `global_move_to`

File: moveos/moveos/src/moveos.rs [hunk 4]
```diff
@@ -175,6 +193,144 @@ impl MoveOS {
...
+            if !self
+                .check_module_init_permission(
+                    session,
+                    tx_context,
+                    &module_id,
+                    INIT_FN_NAME,
+                    vec![],
+                    vec![],
+                )
+                .unwrap()
```
Comment (jolestar) [inline moveos/moveos/src/moveos.rs:278]: do not `unwrap` at here

File: moveos/moveos/src/moveos.rs [hunk 5]
```diff
@@ -175,6 +193,144 @@ impl MoveOS {
...
+            module_id,
+            function_name,
+            ty_args,
+            args,
+            gas_meter,
+        )?;
+        assert!(
+            result.return_values.is_empty(),
+            "Entry function should not return values"
+        );
```
Comment (jolestar) [inline moveos/moveos/src/moveos.rs:330]: This is not an entry function that can return values.

**Review by wow-sven — APPROVED** [review]

**Review by jolestar — COMMENTED** [review]
Add a TODO issue to check the module compatibility and ensure the `init` function only call on the first deployment.

File: moveos/moveos/src/moveos.rs (hunk 3 above)
Comment (baichuan3) [reply moveos/moveos/src/moveos.rs:213]: `init` function doesn't support `signer`, so can not `global_move_to `

File: moveos/moveos/src/moveos.rs [hunk 6]
```diff
@@ -209,12 +249,100 @@ impl MoveOS {
...
+                vec![],
+                vec![],
+            )? {
+                continue;
+            };
+
+            let function =
+                Function::new(module_id, INIT_FN_NAME_IDENTIFIER.clone(), vec![], vec![]);
+            let _result =
+                Self::execute_function_bypass_visibility(session, tx_context, gas_meter, function)?;
```
[the remark this answers is not in the document]
Comment (baichuan3) [reply moveos/moveos/src/moveos.rs:334]: > I think it's better to break immediately once found the init function.

support multi module init

**Review by jolestar — APPROVED** [review]

**Pull request merged**